### PR TITLE
Enable service mirroring to work in private networks

### LINKED ISSUE #4411: service-mirror fails when remote gateway is on a private IP
When service-mirror is configured with a remote gateway on a private (10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16) network, the service-mirror's proxy attempts discovery and the destination service  responds with no identity; and the outbound requests fail because of the `l5d-require-id` header, which requires that the identity is configured (and it is not).

To fix this, I propose the following:
- We create a service for each gateway, e.g. `gateway-multi1`, `gateway-multi2`. This service has two ports (i propose gateway:4180 and healthcheck:4181) and one endpoint for each remote gateway replica.
- The service-mirror calls gateways by their name and not IP.
- The service-mirror manages endpoint readiness according to probes.
- The destination service is able to discover gateway identities (via annotation)?.

wdyt @zaharidichev @grampelberg

### PR DESCRIPTION
**Pull request by zaharidichev**
This change creates a gateway proxy for every gateway. This enables the probe worker to leverage the destination service functionality in order to discover the identity of the gateway. 

Fix #4411

### CODE REVIEW

File: charts/linkerd2-service-mirror/templates/service-mirror.yaml [hunk 1]
```diff
@@ -16,7 +16,7 @@ rules:
   verbs: ["list", "get", "watch", "create", "delete", "update"]
 - apiGroups: [""]
   resources: ["namespaces"]
-  verbs: ["list", "get", "watch"]
+  verbs: ["create","list", "get", "watch"]
```
Comment (adleong) [inline charts/linkerd2-service-mirror/templates/service-mirror.yaml:19]: just to clarify, this is unrelated to this change right?  we would need namespace create permission to create mirror services if their namespace doesn't exist but we won't need it to create the gateway mirrors since those will always be in the service-mirror namespace?

File: cli/cmd/cluster.go [hunk 2]
```diff
@@ -169,8 +169,8 @@ func newGatewaysCommand() *cobra.Command {
 		},
 	}
 
-	cmd.Flags().StringVar(&opts.clusterName, "cluster-name", "remote", "the name of the remote cluster")
-	cmd.Flags().StringVar(&opts.gatewayNamespace, "gateway-namespace", defaultMulticlusterNamespace, "the namespace in which the gateway resides on the remote cluster")
+	cmd.Flags().StringVar(&opts.clusterName, "cluster-name", "", "the name of the remote cluster")
```
Comment (adleong) [inline cli/cmd/cluster.go:172]: just to clarify, leaving these blank means that the default is to look at all remote clusters?  and all gateway namespaces?

File: controller/cmd/service-mirror/cluster_watcher.go [hunk 3]
```diff
@@ -497,6 +500,132 @@ func remapRemoteServicePorts(ports []corev1.ServicePort) []corev1.ServicePort {
 	return newPorts
 }
 
+func (rcsw *RemoteClusterServiceWatcher) updateGatewayProxyService(spec *GatewaySpec) error {
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher.go:503]: calling these gateway proxies is pretty confusing to me.  can we call them "gateway mirrors" or something?

File: controller/cmd/service-mirror/cluster_watcher.go [hunk 4]
```diff
@@ -497,6 +500,132 @@ func remapRemoteServicePorts(ports []corev1.ServicePort) []corev1.ServicePort {
...
+	localSvcProxyName := rcsw.mirroredResourceName(spec.gatewayName)
+	service, err := rcsw.localAPIClient.Svc().Lister().Services(rcsw.serviceMirrorNamespace).Get(localSvcProxyName)
+	if err != nil {
+		return err
+	}
+
+	if service.Annotations != nil && service.Annotations[consts.RemoteGatewayResourceVersionAnnotation] != spec.resourceVersion {
+		updatedService := service.DeepCopy()
+		if updatedService.Annotations != nil {
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher.go:512]: probably don't nee this check because we just checked this in the original service and haven't modified the copy yet.

File: controller/cmd/service-mirror/cluster_watcher.go [hunk 5]
```diff
@@ -497,6 +500,132 @@ func remapRemoteServicePorts(ports []corev1.ServicePort) []corev1.ServicePort {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      localSvcProxyName,
+					Namespace: rcsw.serviceMirrorNamespace,
+					Labels: map[string]string{
+						consts.MirroredResourceLabel:  "true",
+						consts.RemoteClusterNameLabel: rcsw.clusterName,
+					},
+					Annotations: map[string]string{
+						consts.RemoteGatewayIdentity:                  spec.identity,
+						consts.RemoteGatewayResourceVersionAnnotation: spec.resourceVersion,
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher.go:589]: should this annotation be on the service instead of the endpoints?

File: controller/cmd/service-mirror/cluster_watcher.go [hunk 6]
```diff
@@ -497,6 +500,132 @@ func remapRemoteServicePorts(ports []corev1.ServicePort) []corev1.ServicePort {
...
+
+	_, err := rcsw.localAPIClient.Svc().Lister().Services(rcsw.serviceMirrorNamespace).Get(localSvcProxyName)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			serviceToCreate := &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      localSvcProxyName,
+					Namespace: rcsw.serviceMirrorNamespace,
+					Labels: map[string]string{
+						consts.MirroredResourceLabel:  "true",
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher.go:565]: I'm curious what we use this label for, if anything?  I ask because we now have two different kind of services that we create: mirrored services and gateway mirrors.  Anywhere that we're using this label selector will now select both kinds of services.  Is that appropriate?

File: controller/cmd/service-mirror/cluster_watcher.go [hunk 7]
```diff
@@ -585,6 +718,11 @@ func (rcsw *RemoteClusterServiceWatcher) handleRemoteGatewayDeleted(ev *RemoteGa
 		gatewayNs:   ev.gatewayData.Namespace,
 		clusterName: rcsw.clusterName,
 	})
+
+	if err := rcsw.localAPIClient.Client.CoreV1().Services(rcsw.serviceMirrorNamespace).Delete(rcsw.mirroredResourceName(ev.gatewayData.Name), &metav1.DeleteOptions{}); err != nil {
+		rcsw.log.Errorf("Could not delete  gateway proxy %s", err)
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher.go:723]: extra whitespace in this string 🦅 👁️

File: controller/cmd/service-mirror/cluster_watcher.go (hunk 3 above)
Comment (olix0r) [reply controller/cmd/service-mirror/cluster_watcher.go:503]: Better yet, add a comment that clarifies what it actually is?

File: controller/cmd/service-mirror/cluster_watcher.go (hunk 5 above)
Comment (olix0r) [reply controller/cmd/service-mirror/cluster_watcher.go:589]: identities are endpoint-specific and not service-specific, unless we've special-cased this in destination?

Comment (adleong) [reply controller/cmd/service-mirror/cluster_watcher.go:589]: sorry, I mean the resource version annotation.

File: charts/linkerd2-service-mirror/templates/service-mirror.yaml (hunk 1 above)
Comment (zaharidichev) [reply charts/linkerd2-service-mirror/templates/service-mirror.yaml:19]: Yes this is correct

File: cli/cmd/cluster.go (hunk 2 above)
Comment (zaharidichev) [reply cli/cmd/cluster.go:172]: Yes, this fixes the problem you originally mentioned where gateways were not showing up

File: controller/cmd/service-mirror/cluster_watcher.go (hunk 6 above)
Comment (zaharidichev) [reply controller/cmd/service-mirror/cluster_watcher.go:565]: So, we use it in `cleanupMirroredResources` where we combine it with the cluster name. This deletes all mirrored services that needs to be deleted when we stop watching a cluster. We also use it in `mirroredServicesForGateway` but the re we combine it with some more specific annotations so there will be no conflict
